api/envoy/config/bootstrap/v3/bootstrap.proto

```diff
@@ -342,6 +342,15 @@ message Bootstrap {
   // TODO(abeyad): Add public-facing documentation.
   // [#not-implemented-hide:]
   core.v3.TypedExtensionConfig xds_delegate_extension = 35;
+
+  // There are two behaviors can be chosen when update the `socket_options` on the existing listener,
+  // the default behavior is
+  // creating a new socket when the socket_options updated. Another behavior is clone the existing socket
+  // when the socket_options updated, and only the socket option which is set at `STATE_LISTENING` state
+  // can be updated. Whether a socket option can be updated, it depends on the platform. Envoy doesn't
+  // have knowledge about that. The user should ensure the socket update behavior on the platform first,
+  // ensure it works as expected.
+  bool clone_socket_when_options_updated = 36;
```

Comment:
Just from an API perspective, do we need a configuration option for this? Couldn't we do something like:
1) Find the difference in socket options during an update.
2) Based on the difference, ask the platform whether they can be updated or not. The default answer can be no, in which case the config update is NACKd to make it more clear what is happening. Then known safe options or sets of options can be added later?

What happens today? Are the updates just always dropped because we can't do the updates? Or does it depend on reuse_port vs. not and what the exact options are? I think some summary of the current state would be helpful. Thank you!

Reply:
> Just from an API perspective, do we need a configuration option for this? Couldn't we do something like:

If we are going to only support Linux platform, then I prefer to not add this config option, and we just always create new socket and apply the new options. And that avoids adding a lot of complexity to the code.

But as the doc said https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/listener/v3/listener.proto#config-listener-v3-listener, the Linux platform is the only platform that work well with reuse port. If not set reuse port, we aren't able to create a new socket since the port bind conflict.

If we are going to support all the platform, then clone is the only choice can work on the all platforms. Probably we needn't add this config option also, since the create new socket has a downside as you comment in the code:
https://github.com/envoyproxy/envoy/blob/ac5792e1da4c966656f4afc7a184fb39644f24fb/source/server/listener_impl.cc#L128, then clone will become the better option here.

> 
> 1. Find the difference in socket options during an update.

For creating new socket, we needn't do that, we can just apply all the socket_options in the new listener config directly. That is pretty simple.

For clone the existing socket, I think there are two options: First option, we could find out the difference in the socket options. but it is more complex, like the user could remove a socket option, then we should set this option back to the default value. Then another problem shows up, that is collecting all the default value for all the option on the all the platform.  Second option is more simple, we just apply all the new socket options on the cloned socket, asking the user explicitly set an option back to the default value when the user wants to remove an option. of course, this is based on the all socket options update is idempotent, although I didn't see any doc clear about it, but I think why it isn't

> 2. Based on the difference, ask the platform whether they can be updated or not. The default answer can be no, in which case the config update is NACKd to make it more clear what is happening. Then known safe options or sets of options can be added later?

For creating new socket, we needn't care about this. Since it just applies all socket options in the new listener config.

For clone existing socket, we probably needed a rollback mechanism, if the platform reject the update, then we rollback. Or we also can leave the responsibility to the user, asking the user test the socket option update behavior on the platform first.

> 
> What happens today? Are the updates just always dropped because we can't do the updates? Or does it depend on reuse_port vs. not and what the exact options are? I think some summary of the current state would be helpful. Thank you!

Today, we will clone(duplicate) the existing socket when update listener, the socket options are cloned from the old listener also. Then the change in the new listener config is just ignored.
https://github.com/envoyproxy/envoy/blob/ac5792e1da4c966656f4afc7a184fb39644f24fb/source/server/listener_impl.cc#L122

To summarize the solution here:
* only support creating a new socket option. The implementation is simple. The downside is only supported Linux platform, does not support reuse_port is false. And https://github.com/envoyproxy/envoy/blob/ac5792e1da4c966656f4afc7a184fb39644f24fb/source/server/listener_impl.cc#L128
* only support clone existing socket option. support all the platforms, and support both the resuse_port is true and false. The downside is adding a lot of complexity to the code. But we could reduce the complexity by dropping some responsibilities if that are acceptable.

Reply:
Thanks for the detailed explanation. IMO this is super complicated and probably not worth the effort to do something even more complicated. I would probably:
1) Only support changing socket options when reuse_port is true, so effectively only on Linux currently.
2) NACK the config update in all other situations of the socket options change so it's more clear what is going on.

Can we start there?

Reply:
thanks! yes, we can start here, that is what I prefer too.